### PR TITLE
[41] As a user, I can use advanced search queries to search through reports

### LINKED ISSUE #41: As a user, I can use advanced search queries to search through reports
## Why

As part of the requirements, the user should be able to query for search results with SQLesk queries.

## Acceptance Criteria

1. The user can use valid wildcard options for filtering search results.
2. Only selected queries will be allowed and create queries dynamically.

### PR DESCRIPTION
**Pull request by bterone**
Closes #41

## What happened

✅: Add advanced search filter

We can filter by using the following by typing into the search bar with operations, similar to clubhouse or github.

Here are some examples
```
# Normal Filtering
coffee title:beverages url:drinks.com top_ads:<20 regular_ads:>25

# Query with filter in the middle of string
random title:"Report Title" item 

# Filtering with double quotes
vacation hotspots title:"Business & Travel" url:"https://www.travel.com"
```

## Insight

When the user inputs their query into the search bar,
We're first parsing that query into a params format 
`E.g. Random String title:"Report Title" => %{search_term: "Random String", title: "Report Title"}`

From there, we validate the params though a casted changeset and pass it through to the search_results context for querying and filtering.

## Proof Of Work

For the search term SQL and only in the report with the title "test1"
Screen Shot 2021-06-08 at 12 05 04 PM

For report title with the name "test1" and less than 4 top_ads
Screen Shot 2021-06-08 at 12 07 29 PM

For report title with the name "test1" and exactly 3 top_ads
Screen Shot 2021-06-08 at 12 07 41 PM

### CODE REVIEW

File: lib/g_searcher_web/controllers/error_handler.ex [hunk 1]
```diff
@@ -10,4 +12,20 @@ defmodule GSearcherWeb.ErrorHandler do
     |> render("#{status}_page.html")
     |> halt()
   end
+
+  def combine_changeset_errors(changeset) do
```
Comment (hanam1ni) [inline lib/g_searcher_web/controllers/error_handler.ex]: How about rename this function to be more related to building error messages? I confuse read function name at first.

File: lib/g_searcher_web/controllers/helpers/search_helper.ex [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      |> convert_search_options_to_map()
+      |> build_query_attributes()
+
+    {:ok, parsed_query}
+  end
+
+  def parse_query(nil), do: {:error, :no_query}
+
+  defp split_string_by_space(query, ignore_in_double_quotes: true) when is_binary(query),
+    do: Regex.split(~r{\s+(?=([^"]*"[^"]*")*[^"]*$)}, query)
```
Comment (hanam1ni) [inline lib/g_searcher_web/controllers/helpers/search_helper.ex]: How about moving this to the module attribute? with the name, it easier to understand.

File: lib/g_searcher_web/controllers/helpers/search_helper.ex [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    end)
+  end
+
+  defp join_non_operator_filters(list) do
+    list
+    |> Enum.reject(&is_map/1)
+    |> Enum.join(" ")
+  end
+
+  defp escape_percentage_sign(string), do: String.replace(string, "%", "\\%")
```
Comment (hanam1ni) [inline lib/g_searcher_web/controllers/helpers/search_helper.ex]: Let's reorder the function by the order its called in this module.

File: test/g_searcher_web/controllers/helpers/search_helper_test.exs [hunk 4]
```diff
@@ -0,0 +1,54 @@
+defmodule GSearcherWeb.Helpers.SearchHelperTest do
+  use GSearcherWeb.ConnCase
```
Comment (hanam1ni) [inline test/g_searcher_web/controllers/helpers/search_helper_test.exs]:
```suggestion
  use GSearcherWeb.ConnCase, aync: true
```

File: lib/g_searcher_web/controllers/helpers/search_helper.ex [hunk 5]
```diff
@@ -0,0 +1,60 @@
+defmodule GSearcherWeb.Helpers.SearchHelper do
+  @valid_operators ["title", "url", "top_ads", "regular_ads", "status"]
+
+  @spec parse_query(String.t()) :: {:ok, %{String.t() => String.t()}}
+  def parse_query(query) when is_binary(query) do
+    parsed_query =
+      query
+      |> escape_percentage_sign()
+      |> split_string_by_space(ignore_in_double_quotes: true)
```
Comment (hanam1ni) [inline lib/g_searcher_web/controllers/helpers/search_helper.ex]: Will we using `ignore_in_double_quotes` option in the future? 🤔

File: lib/g_searcher_web/controllers/search_result_controller.ex [hunk 6]
```diff
@@ -3,13 +3,31 @@ defmodule GSearcherWeb.SearchResultController do
...
     %{id: user_id} = conn.assigns.current_user
 
-    search_results = SearchResults.list_search_results_by_user_id(user_id, params["query"])
+    with {:ok, query_params} <- SearchHelper.parse_query(params["query"]),
+         {:ok, validated_params} <-
+           ParamsValidator.validate(query_params, as: SearchResultParams),
+         search_results <- SearchResults.list_search_results_by_user_id(user_id, validated_params) do
+      render(conn, "index.html", search_results: search_results)
+    else
+      {:error, :no_query} ->
```
Comment (hanam1ni) [inline lib/g_searcher_web/controllers/search_result_controller.ex]: It is weird to put the path that is not error handling here. How about use pattern matching for `index` action?
```elixir
def index(conn, %{"query" => query}) when is_binary(query) do
  # Validate params and do a search
end

def index(conn, params) do
  %{id: user_id} = conn.assigns.current_user

  search_results = SearchResults.list_search_results_by_user_id(user_id)        
  
  render(conn, "index.html", search_results: search_results)
end
```

File: lib/g_searcher_web/controllers/helpers/search_helper.ex (hunk 5 above)
Comment (bterone) [reply lib/g_searcher_web/controllers/helpers/search_helper.ex]: I was using it to be a bit more descriptive. I think the module attribute for the regex makes it's descriptive enough. 👍

File: lib/g_searcher_web/controllers/search_result_controller.ex (hunk 6 above)
Comment (bterone) [reply lib/g_searcher_web/controllers/search_result_controller.ex]: That's a great idea! Thanks 😄 .
I've fixed it in 0349945a85fcf825a9e941873e03ea6059c4ce49

**Review by hanam1ni — APPROVED** [review]
LGTM

**Comment by andyduong1920** [comment]
LGTM @bterone  👍 💪

**Pull request merged**